cpp/src/arrow/types/struct-test.cc

```diff
@@ -52,4 +62,133 @@ TEST(TestStructType, Basics) {
   // TODO(wesm): out of bounds for field(...)
 }
 
-}  // namespace arrow
+// .............................................................................
+// Struct test
+class TestStructBuilder : public TestBuilder {
+ public:
+  void SetUp() {
+    TestBuilder::SetUp();
+
+    auto value_type = TypePtr(new Int32Type());
+    auto char_type = TypePtr(new Int8Type());
+    auto list_type = TypePtr(new ListType(char_type));
+
+    std::vector<TypePtr> types = {list_type, value_type};
+    std::vector<FieldPtr> fields;
+    fields.push_back(FieldPtr(new Field("list", list_type)));
+    fields.push_back(FieldPtr(new Field("int", value_type)));
+
+    type_ = TypePtr(new StructType(fields));
+    value_fields_ = fields;
+
+    std::shared_ptr<ArrayBuilder> tmp;
+    ASSERT_OK(MakeStructBuilder(pool_, type_, fields, &tmp));
+
+    builder_ = std::dynamic_pointer_cast<StructBuilder>(tmp);
+  }
+
+  void Done() {
+    result_ = std::dynamic_pointer_cast<StructArray>(builder_->Finish());
+  }
+
+ protected:
+  std::vector<FieldPtr> value_fields_;
+  TypePtr type_;
+
+  std::shared_ptr<StructBuilder> builder_;
+  std::shared_ptr<StructArray> result_;
+};
+
+TEST_F(TestStructBuilder, TestAppendNull) {
+  ASSERT_OK(builder_->AppendNull());
+  ASSERT_OK(builder_->AppendNull());
+  ASSERT_EQ(2, builder_->value_builder().size());
+
+  Done();
+
+  ASSERT_EQ(2, result_->values().size());
+  ASSERT_EQ(2, result_->length());
+  ASSERT_TRUE(result_->IsNull(0));
+  ASSERT_TRUE(result_->IsNull(1));
+
+
+  auto list_char = static_cast<ListArray*>(result_->values(0).get());
+  auto chars = static_cast<Int8Array*>(list_char->values().get());
+  auto int32 = static_cast<Int32Array*>(result_->values(1).get());
+  ASSERT_EQ(0, list_char->length());
+  ASSERT_EQ(0, chars->length());
+  ASSERT_EQ(0, int32->length());
```

Comment:
I touched on this above. I think these results are malformed -- the child arrays must have length equal to the parent StructArray